Initial People Management System Service Functionality

CiaranJK opened the pull request.

ciaranjkieran committed:
Initial commit

CiaranJK committed:
Add files via upload

GabrielGrimberg commented on main/kotlin/com/example/people_managment_system/controller/SystemController.kt:
```diff
@@ -0,0 +1,52 @@
+
+@RestController
+@RequestMapping("api/system")
+
```
No need for this extra line, it can be removed.

GabrielGrimberg commented on main/kotlin/com/example/people_managment_system/controller/SystemController.kt:
```diff
@@ -0,0 +1,52 @@
+import org.springframework.web.bind.annotation.*
+
+@RestController
+@RequestMapping("api/system")
```
Good work on the controller. This controller has different operations to do with Company, you can split it by creating two controllers.

One can be `EmployeeController` and the other can be `CompanyController`.

GabrielGrimberg commented on main/kotlin/com/example/people_managment_system/PostgresqlApplication.kt:
```diff
@@ -0,0 +1,12 @@
+class PostgreSqlApplication
+
+fun main(args: Array<String>) {
+	val runApplication = runApplication<PostgreSqlApplication>(*args)
```
I think you can have it like: `runApplication<PostgreSqlApplication>(*args)` in which should also work because you are not using this variable anywhere.

GabrielGrimberg commented on main/kotlin/com/example/people_managment_system/PostgresqlApplication.kt:
```diff
@@ -0,0 +1,12 @@
+package com.example.people_managment_system
```
I recommend to avoid using underscores, hyphens or any of these characters for packages.

Also recommend following a company domain to make it a better example, maybe go with `com.ciarankieran.core.service`

`core` indicates the domain. A domain can be like payroll or something specific in a company and `service` would be a part of `core`. If it's authentication then it can be `core.authentication`.

GabrielGrimberg commented on main/kotlin/com/example/people_managment_system/controller/SystemController.kt:
```diff
@@ -0,0 +1,52 @@
+    fun handleNotFound(e: NoSuchElementException): ResponseEntity<String> =
+        ResponseEntity(e.message, HttpStatus.NOT_FOUND)
+
+    @ExceptionHandler(DataIntegrityViolationException::class)
```
Exception handling is always a must and it's good that you added it in here.

What I would recommend is creating a class called `ExceptionHandler` and give it the `@RestControllerAdvice` annotation. 

Think of this class as the source of truth for handling exceptions in your service.

You can create methods to handle exceptions and tell the method what type of a `ResponseEntity` you'd like to return back to the user.

Check this link out for more information: https://codersee.com/exception-handling-with-restcontrolleradvice-and-exceptionhandler/

GabrielGrimberg commented on main/kotlin/com/example/people_managment_system/controller/SystemController.kt:
```diff
@@ -0,0 +1,52 @@
+package com.example.people_managment_system.controller
```
I recommend that whenever you have an API type of class such as in your case is HTTP. Then create a package structure such as: `.../api/http/v1/<Controller Classes>`.

When we create API, we version them because in the future we will look to up their versions with new improvements while supporting older versions for legacy up until everyone has upgraded to the latest version. This is a standard practice to follow because once you release an API, you should not look change the API contract, what I mean by that is changing the API URL, because this would break for the customer using it.

GabrielGrimberg commented on main/kotlin/com/example/people_managment_system/dtos/EmployeeDto.kt:
```diff
@@ -0,0 +1,13 @@
+package com.example.people_managment_system.dtos
```
Instead of `dtos` as the package, I recommend calling it `domain`.

GabrielGrimberg commented on main/kotlin/com/example/people_managment_system/dtos/EmployeeDto.kt:
```diff
@@ -0,0 +1,13 @@
+
+import java.io.Serializable
+
+data class EmployeeDto(
```
No need to have the `Dto` name in the class as we can understand what it is by seeing it is a `data class`.

CiaranJK replied:
There is another class called Employee that corresponds to employee entities in the 'employees' table in the Postgres database so I left this name as it was.  See if you think this was correct when you see the updated code.

GabrielGrimberg commented on main/kotlin/com/example/people_managment_system/dtos/EmployeeDto.kt:
```diff
@@ -0,0 +1,13 @@
+    var email: String? = null,
+    var salary: Int? = null,
+    var companyName: String? = null,
+    var managerID: Int? = null,
```
I recommend to always follow `camelCase` naming convention for variables in Kotlin.

GabrielGrimberg commented on main/kotlin/com/example/people_managment_system/dtos/EmployeeDto.kt:
```diff
@@ -0,0 +1,13 @@
+import java.io.Serializable
+
+data class EmployeeDto(
+    var employeeID: Int? = null,
```
There are many fields in here that can be nullable upon a creation of an employee, I recommend making it more strict to ensure values get persisted for employees.

For example, the `employeId` is a crucial value, if this is null then it can cause problems.

CiaranJK replied:
I changed this, thanks.  It is no longer nullable.  It seems I had to initialise the variable with 0 in order to get the code to compile.  See what you think.

GabrielGrimberg commented on main/kotlin/com/example/people_managment_system/model/Company.kt:
```diff
@@ -0,0 +1,22 @@
+package com.example.people_managment_system.model
+
+import javax.persistence.*
```
You are importing everything, only import what you are using.

CiaranJK replied:
I think I needed to import everything in the module and that's why it's importing everything.  When I imported what I needed to import, the import statements automatically changed to import 'javax.persistence.*'

GabrielGrimberg commented on main/kotlin/com/example/people_managment_system/model/Company.kt:
```diff
@@ -0,0 +1,22 @@
+package com.example.people_managment_system.model
```
These files can go in `domain/model`.

GabrielGrimberg commented on main/kotlin/com/example/people_managment_system/model/Company.kt:
```diff
@@ -0,0 +1,22 @@
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Column(name = "company_id", nullable = false)
+    open var companyId: Int? = null
```
Always think about the data and understand what we always need upon creating something and what can be null and added later.

In this case it seems `companyId` is vital to create this row in the table, so it should not be nullable. You have it as nullable set to false but the code can set it to null by default, this can cause confusions.

CiaranJK replied:
Thanks for this, I changed this in the code.  The company ID is no longer nullable.  But it seems I had to initialise the variable to get the code to compile.  I initialised it as 0 in the class, but this value subsequently changes when a new entity is created in the database.

GabrielGrimberg commented on main/kotlin/com/example/people_managment_system/model/Employee.kt:
```diff
@@ -0,0 +1,41 @@
+
+
+
+
```
You can remove these empty lines.

GabrielGrimberg commented on main/kotlin/com/example/people_managment_system/repositories/CompanyRepository.kt:
```diff
@@ -0,0 +1,12 @@
+package com.example.people_managment_system.repositories
```
Seems like you have four domains in this service: Company, Employee, Manager and Team. 

There many ways to structure the project and it's up to you on how you want to do it. A recommendation is that you have four packages in `domain`. Something like: `domain.company`, `domain,employee`, `domain.manager` and `domain.team`. 

You can also structure it under an `.../infrastructure` package, this one is in the name basically logic or code that executes such as writing to a Database.

GabrielGrimberg commented on main/kotlin/com/example/people_managment_system/service/SystemService.kt:
```diff
@@ -0,0 +1,177 @@
+package com.example.people_managment_system.service
```
Good stuff on this, now this is what we would call the business logic, operations to perform specific tasks such as getting an employee.

Instead of putting this under the `service` package. I'd recommend changing it to `usecases`.

Similar to the controller, this class is too big, it's doing too much. We need to separate it out into different classes.

Let's use the Employee as the example. We want to perform some actions on it. So, we will create a package called `usecases/employee`. Inside this package we will have a class called `EmployeeService`. Here you can put all the operations related only to Employee.

I also recommend implementing to an interface. Meaning the `EmployeeService` class will have a blueprint of what methods it should implement.

You can read more about it here: https://www.geeksforgeeks.org/implement-interface-in-kotlin/

I will let you figure the other classes out.

GabrielGrimberg commented on main/kotlin/com/example/people_managment_system/service/SystemService.kt:
```diff
@@ -0,0 +1,177 @@
+
+    fun retrieveEmployeesInACompany(companyName: String): MutableList<EmployeeDto> {
+    val employeesFound = mutableListOf<EmployeeDto>()
+    for (item in employeeRepository.findAll()) {
```
This is a bonus, but you can take a look into high order functions, basically the modern way to program: https://kotlinlang.org/docs/lambdas.html

GabrielGrimberg commented on main/kotlin/com/example/people_managment_system/service/SystemService.kt:
```diff
@@ -0,0 +1,177 @@
+
+
+
+
```
Can remove the blank lines.

GabrielGrimberg commented on main/kotlin/com/example/people_managment_system/service/SystemService.kt:
```diff
@@ -0,0 +1,177 @@
+           updateManager()
+*/
+   fun mapEmployeeEntityToEmployeeDto(employeeEntity: Employee): EmployeeDto {
+       val employeeDto = EmployeeDto()
```
You can directly put items into the EmployeeDto(..).

Example:

```
return EmployeeDto(
       employeeEntity.employeeID
       employeeEntity.employeeName
       employeeEntity.email
       employeeEntity.salary
       employeeEntity.company.toString()
       employeeEntity.manager?.managerID
       employeeEntity.team?.teamID
)
```

GabrielGrimberg commented on main/kotlin/com/example/people_managment_system/service/SystemService.kt:
```diff
@@ -0,0 +1,177 @@
+/* Deletes an employee's data.
+   Returns a String response if the employee ID does not exist. */
+
+    fun deleteEmployee(relevantEmployeeId: Int): String {
```
When you move this to the EmployeeService class, you can call this method `delete`. Because, we know which class it is and it's a specific operation for Employee.

You can do similar to other classes such as Company.

GabrielGrimberg commented on main/kotlin/com/example/people_managment_system/service/SystemService.kt:
```diff
@@ -0,0 +1,177 @@
+}
+
+/* Deletes an employee's data.
+   Returns a String response if the employee ID does not exist. */
```
Usually with a delete operation, you don't need to return anything, you can pass the ID to delete and just execute it, expect that once the ID is passed and executed, if it's there it's deleted if no exception is thrown and if it's not there then it's deleted anyway.

GabrielGrimberg commented on main/kotlin/com/example/people_managment_system/service/SystemService.kt:
```diff
@@ -0,0 +1,177 @@
+
+@Service
+
+class SystemService(
```
When you commend code, you should always look to comment public classes and methods.

You can use KDoc for this.

It looks like this:

```
/***
* ...
* ...
**
```

GabrielGrimberg commented on main/resources/application.properties:
```diff
@@ -0,0 +1,5 @@
+spring.datasource.url=jdbc:postgresql://localhost/people_ms
+spring.datasource.username=postgres
+spring.datasource.password=sAU-uf,S5W=^/)Yu
```
You should never ever add actual password to properties. Just a thought, it's okay here because this is an example but in real world applications, this must always be encrypted.

CiaranJK replied:
Thanks Gabz, I changed this by encrypting the password with Jasypt and storing it in a file called 'encrypted.properties' which I won't be sharing on GitHub.

GabrielGrimberg commented on test/kotlin/com/example/people_managment_system/PostgresqlApplicationTests.kt:
```diff
@@ -0,0 +1,13 @@
+import org.springframework.boot.test.context.SpringBootTest
+
+@SpringBootTest
+class PostgresqlApplicationTests {
```
If a test has nothing in it then it can be deleted.

GabrielGrimberg commented on main/resources/static/index.html:
```diff
@@ -0,0 +1 @@
+<h1>Hello world</h1>
```
Can be deleted if not used.

GabrielGrimberg commented on test/kotlin/com/example/people_managment_system/controller/SystemControllerTest.kt:
```diff
@@ -0,0 +1,308 @@
+@SpringBootTest
+@AutoConfigureMockMvc
+@ContextConfiguration
+internal class SystemControllerTest @Autowired constructor (
```
It's always good to have test coverage for the functionality you implement, this one is a controller test and you should also look to make tests for the SystemService class (this will be four classes or as many needed depending on the functionary you have as discussed in the previous comment). 

You don't need to inject dependencies into this class, most of the time you don't but here no need.

You can have it like.

```
@WebMvcTest(controllers = [SystemController::class])
@ActiveProfiles("test")
internal class SystemControllerTest {

    @Autowired
    private lateinit var mockMvc: MockMvc

}
``` 

ObjectMapper is not a dependency for the class so I'm not sure why it's injected.

CiaranJK replied:
I tried this but couldn't get it to compile so I decided to leave it with the original format for now.  

I think the ObjectMapper dependency is used in the Controller Test classes when I am mapping Data Transfer Objects as strings when testing post requests. 

Have a look and see what you think.  It may be that I am not using it right.

GabrielGrimberg commented on test/kotlin/com/example/people_managment_system/controller/SystemControllerTest.kt:
```diff
@@ -0,0 +1,308 @@
+    @DisplayName("getEmployee()")
+    @TestInstance(TestInstance.Lifecycle.PER_CLASS)
+
+    inner class GetEmployee {
```
This is one way of doing it but there is another way and this is basically mocking the SystemService class.

You can check this guide out in how to mock with mockito: https://www.baeldung.com/kotlin/mockito

GabrielGrimberg commented:
Good stuff on this project Ciaran, it's a good start into learning Spring Boot. I have left some comments for the first iteration of this project, once changes are made, I will leave more comments and we can look to improve this project further.

A tip, when writing up a merge request, ensure there is a description in there to state the changes you have done, this will allow the reviewer to understand more about is happening. For this case it's fine but when working in a team this is needed.

GabrielGrimberg reviewed (changes requested):
Good stuff on this project Ciaran, it's a good start into learning Spring Boot. I have left some comments for the first iteration of this project, once changes are made, I will leave more comments and we can look to improve this project further.

A tip, when writing up a merge request, ensure there is a description in there to state the changes you have done, this will allow the reviewer to understand more about is happening. For this case it's fine but when working in a team this is needed. A meaningful title would also be good. For this merge request you can change the title from Review1 to something like **Initial People Management Service Functionality**.

Also, change the destination of this merge request. Instead of `src`, use `master`.

ciaranjkieran committed:
Implemented changes following:
Initial People Management System Service Functionality Pull Request

CiaranJK committed:
Add files via upload

CiaranJK committed:
Add files via upload

CiaranJK committed:
Delete src directory

CiaranJK committed:
Delete src directory

ciaranjkieran committed:
Delete src directory

ciaranjkieran committed:
Delete main directory

ciaranjkieran committed:
Delete test/kotlin/com/example/people_managment_system directory

ciaranjkieran committed:
Add files via upload

ciaranjkieran committed:
Delete encrypted.properties

ciaranjkieran committed:
Delete src directory

ciaranjkieran committed:
Add files via upload

ciaranjkieran committed:
Add files via upload

CiaranJK commented:
Great feedback Gabriel.  I very much enjoyed updating the project.  See what you think of the new version.  I changed the structure a lot in accordance with your instructions.  I split up the domains as you suggested, and divided up the service and controller functionality into 4 different classes.  I also added some implementation of interfaces and some extra functionality for each of the 4 domains - Company, Employee, Manager and Team.

Also testing for all of the different methods in the project.

I experimented with different class structures that incorporated inheritance and polymorphism but I decided to go with the structure it has now.
